docs/modules/sql/pages/user-defined-types.adoc

```diff
@@ -1,60 +1,53 @@
 = User-Defined Types in SQL
 
-User-Defined Types (also known as UDTs, nested types, or nested fields) is an experimental feature that allows you to create
- custom data types that can be referenced in the xref:sql:create-mapping.adoc[CREATE MAPPING statement].
+User-Defined Types (also known as UDTs, nested types, or nested fields) allow you to create custom data types that can be referenced in the xref:sql:create-mapping.adoc[CREATE MAPPING statement].
 
-UDTs are useful for creating and accessing hierarchical data structures, including simple cases of nested objects and
-more complex cases of fully/partially connected graphs of objects. See <<queryingExamples,examples>> for more information.
-
-WARNING: Support for UDTs is an experimental feature, and disabled by default.
-See <<enabling-udt-support, Enabling UDT support>> to learn how to enable it.
+UDTs are useful for creating and accessing hierarchical data structures, which can even be cyclic. See <<queryingExamples,examples>> for more information.
 
 == Feature Overview
 
-Due to the experimental nature of UDTs, the feature set is limited to the following:
+- All serialization formats (namely `portable`, `compact`, `java` and `avro`) are supported.
+- Cyclic definitions are only supported for Java-serialized structures. Portable and Compact formats do not support cyclic schemas, and Avro support is currently limited to acyclic schemas.
++
+NOTE: If mappings use cyclic structures, the `INSERT` and `UPDATE` queries are disabled.
+- Type options override mapping options. Where both the mapping and type define a schema (Portable class definition, Compact schema, Java class or Avro schema) and the mapping contains only `__key` or `this` fields, the schema defined by the type is used. This means that the mapping does not need to define a schema if the type defines one.
+- If the type does not specify a schema, the parent structure is used. The parent structure could be a mapping or another type.
+- Type fields are optional and are resolved from the schema if not specified. Following a `CREATE MAPPING` statement, type fields are resolved and materialized. You can check the resolved fields using a `GET_DDL` query after the type has been used in a mapping. If type fields are not specified:
```

Comment:
AFAIK this "magic" materialization was a side effect and not something uses should rely on, as it is subject to change in the future

Reply:
and were not sure if it makes sense in the long term

Reply:
I removed it from notes. Now, it is only mentioned in "Replacing Types and Type Consistency". (099f756d)

Reply:
I do not see the place you mentioned. However I can see:

> If the type does not specify a schema, it is resolved from the parent structure, which may be a mapping or another type.

(and next bullet) which makes this magic UDT field resolution a "normal" feature, but we had doubts if we want to keep it

Reply:
Field materialization is mentioned in "Replacing Types and Type Consistency" as the following, and I think this is the right place for such an implementation detail.
> ... if the type hierarchy was not used in a mapping, any type in that hierarchy can be safely changed, and these changes will appear in the new mapping. **This is because the links between types are symbolic (based on the name only), and they are materialized only when used in a mapping.**

Reply:
this sentence is about UDT references ("links between types") not fields of fieldless UDT. It is not obvious that is it the same.

And this is not only implementation detail. It has visible impact on the user. We say that:

> Type fields are optional and if not specified, they are resolved from the class/schema when the type is used in a CREATE MAPPING statement.

but we do not say what happens if the underlying schema changes (eg. Java class is updated).
On it's own the above sentence could mean that fieldless UDT may get different set of fields for each context in which it used which is not the way we work.

Reply:
7bc63e25